app/views/course/assessment/answer/programming/_test_cases.html.slim

```diff
@@ -1,127 +1,79 @@
+            th = Course::Assessment::Question::ProgrammingTestCase.human_attribute_name(:expression)
+            th = Course::Assessment::Question::ProgrammingTestCase.human_attribute_name(:expected)
+            - if is_grader
+              th = t('.output')
             - if @submission.attempting?
+              th = Course::Assessment::Question::ProgrammingTestCase.human_attribute_name(:hint)
+            th = t('.passed')
+        tbody
+          tr
+          - test_cases_by_type[test_case_type].each do |test_case|
+            - test_case_result = answer_test_results_hash[test_case]
+            = content_tag_for(:tr, test_case) do
+              - if is_grader
+                th = format_html(test_case.identifier)
+              td = format_html(test_case.expression)
+              td
+                div.expected = format_html(test_case.expected)
+              - if is_grader
+                td = simple_format(get_output(test_case_result))
+              - if @submission.attempting?
+                td
+                  - if !test_case_result || !test_case_result.passed?
+                    = simple_format(get_hint(test_case, test_case_result))
               td
-                - if !test_case_result || !test_case_result.passed?
-                  = simple_format(get_hint(test_case, test_case_result))
-            td
-              - if test_case_result && test_case_result.passed?
-                = fa_icon 'check'.freeze
-              - elsif test_case_result
-                = fa_icon 'times'.freeze
+                - if test_case_result && test_case_result.passed?
+                  = fa_icon 'check'.freeze
+                - elsif test_case_result
+                  = fa_icon 'times'.freeze
 
-- failed_public_test_cases = public_test_cases. \
-    select { |t| answer_test_results_hash[t] && !answer_test_results_hash[t].passed? }
-- if !failed_public_test_cases.empty?
-    - test_case = failed_public_test_cases.first
-    - test_case_result = answer_test_results_hash[test_case]
-    div.panel.panel-danger
-      div.panel-heading = t('.failed_public_message')
-      div.panel-body = simple_format(get_hint(test_case, test_case_result))
+/ If there are no public tests, private hints should still show
+- passed_all_public_tests = true
 
-- failed_private_test_cases = private_test_cases. \
-    select { |t| answer_test_results_hash[t] && !answer_test_results_hash[t].passed? }
-- if failed_public_test_cases.empty? && !failed_private_test_cases.empty?
-    - test_case = failed_private_test_cases.first
-    - test_case_result = answer_test_results_hash[test_case]
-    div.panel.panel-danger
-      div.panel-heading = t('.failed_private_message')
-      div.panel-body = simple_format(get_hint(test_case, test_case_result))
+/ Display warning panels with the first hint when there are failing tests
+- test_case_types.each do |test_case_type|
```

Comment:
I will read through in more detail tomorrow, but I suppose we can move some of these logic into helpers.